proof/refine/ARM/KHeap_R.thy

```diff
@@ -3314,5 +3348,86 @@ lemma refillSufficient_wp:
   apply (wpsimp wp: getRefills_wp)
   by (clarsimp simp: sufficientRefills_def obj_at'_def)
 
+lemma update_sched_context_rewrite:
+  "monadic_rewrite False True (sc_obj_at n scp)
+    (update_sched_context scp f)
+      (do sc \<leftarrow> get_sched_context scp;
+          set_object scp (kernel_object.SchedContext (f sc) n) od)"
+  apply (clarsimp simp: update_sched_context_def get_sched_context_def bind_assoc)
+  apply (rule monadic_rewrite_bind_tail)
+   defer
+   apply (rule get_object_sp)
+  apply (case_tac obj; clarsimp simp: monadic_rewrite_refl3 set_object_def)
+  apply (rule monadic_rewrite_bind_tail)
+   defer
+   apply (rule get_object_sp)
+  apply (clarsimp simp: monadic_rewrite_def obj_at_def is_sc_obj_def)
+  done
+
+lemma get_sched_context_exs_valid:
+  "\<exists>sc n. kheap s scp = Some (Structures_A.SchedContext sc n)
+   \<Longrightarrow> \<lbrace>(=) s\<rbrace> get_sched_context scp \<exists>\<lbrace>\<lambda>_. (=) s\<rbrace>"
+  by (clarsimp simp: get_sched_context_def get_object_def obj_at_def bind_def is_sc_obj_def
+                     gets_def get_def return_def exs_valid_def
+              split: Structures_A.kernel_object.splits)
+
+lemma get_sched_context_no_fail:
+  "no_fail (\<lambda>s. sc_at ptr s) (get_sched_context ptr)"
+  by (clarsimp simp: get_sched_context_def no_fail_def bind_def get_object_def return_def get_def
+                     gets_def obj_at_def is_sc_obj_def
+              split: Structures_A.kernel_object.splits)
+
+(* this let us cross the sc size information from concrete to abstract *)
+lemma ko_at'_cross:
+  assumes p: "pspace_relation (kheap s) (ksPSpace s')"
+  assumes t: "ko_at' (sc'::sched_context) ptr s'"
+  shows "sc_obj_at (objBits sc' - minSchedContextBits) ptr s" using assms
+  apply (clarsimp simp: obj_at'_def projectKOs)
+  apply (erule (1) pspace_dom_relatedE)
+  by (clarsimp simp: obj_relation_cuts_def2 obj_at_def is_sc_obj cte_relation_def
+                     other_obj_relation_def pte_relation_def pde_relation_def
+                     scBits_simps sc_relation_def objBits_simps
+              split: Structures_A.kernel_object.split_asm if_split_asm
+                     ARM_A.arch_kernel_obj.split_asm)
+
+lemma update_sc_no_reply_stack_update_corres:
```

Comment:
This proof is quite a journey in that it goes:
`sc_at` -> `sc_at’` -> symb_exec of `getSchedContext` to fix `sc’` -> `ko_at’` -> `sc_at_n`

It makes sense that this works. But I wonder if there is a more direct way to do something like this.